### ISSUE TITLE
Create a definition of done

### ISSUE THREAD

**Issue by sktran** [boltlabs-inc/key-mgmt#5]
Establishing a "definition of done" creates transparency for project contributors and builds a shared understanding of the conditions that must be satisfied before any work can be considered done. Such a definition sets the standard for functionality and quality for the team and helps individual contributors close out tasks once all necessary elements are completed.

It may be helpful to think of the definition of done as the minimum work required to meet the team's agreed-upon quality level, and it serves as a checklist for reviewers to surface any issues or questions.

For this project, a task will be considered to be complete when the below items have been carried out.

> 1. The task is described in a GitHub issue, which states the problem that it is solving and outlines the acceptance criteria for the work.
> 2. The code is readable and self-explanatory for the reader, and there are comments where appropriate to provide clarity.
> 3. Documentation and diagrams are produced, including relevant updates to the specification.
> 4. Public APIs are properly documented. (Public can be internal as well. If the API needs to be called from within another program, it needs to be documented as well.)
> 5. All "checks" pass, ideally we also have a GitHub action that verifies this. Checks include: formatting (e.g. Cargo fmt), linting (e.g. Clippy), all tests run, and so on.
> 6. The new code has testing infrastructure: unit tests are fully implemented. Integration tests and other testing infrastructure is planned: either fully implemented or via new issues describing the plan for creating them. Also edge cases need to be covered.
> 7. The code has been reviewed by at least one person other than the initial writer, either by pair programming or by review. The reviewer is responsible for verifying the “Definition of Done” before accepting.
> 8. The issue is closed and any related pull request has been merged.
> 9. All commits/PRs that were made related to this issue are properly linked (i.e., include the issue number in commit messages for traceability).

The above is slightly modified from an existing proposal. What can we clarify to make this definition for streamlined?

To close out this issue, comments will be merged into an agreed-upon definition of done for this project. The finalized version will be stored in a `definition-of-done.md` file within the repo.

**Comment by sktran** [comment]
I would especially like @gijsvl and @marsella to identify any items that require better clarification based on past development experiences.

I would also like @indomitableSwan to clarify any language around work on the specification. I know this was something you wanted to include in the definition.

It will be helpful to be as a specific as possible to allow reviewers to navigate through the definition of done during code reviews. (But we don't want this definition to be bloated with too many steps!)

**Comment by indomitableSwan** [comment]
I think it would be helpful to move away from the umbrella term "task" and break down what we mean more precisely. It seems like we have dual goals: to establish guidelines for internal contributions/development, and to define what "done" means for different objects/artifacts. 

It seems we have at least the following objects:
-- issues: What makes a good issue description? I think (1) starts to address this, but could be expanded somewhat: maybe issues should link to any relevant sections of the specification; if we want to use effort estimates, tags, etc, we need to talk about how to do so effectively here. what are good acceptance criteria?
-- pull requests: When is a PR ready for review? (e.g., code meets development guidelines, elements from (1), (3),  (5), (9))

And the following development processes/ coding standards:
- What are our development guidelines? What should programmers keep in mind while they are developing? e.g. (2), (3), (4), (6) all touch on this area.
- What process do we expect reviewers to follow in evaluating PRs? Does (7) cover everything here?
- Whose job is it to merge PRs, once approved?
